Cilsil/Cil/Parsers/InstructionParser.cs

```diff
@@ -553,18 +573,73 @@ protected void RegisterNode(ProgramState state, CfgNode node)
                                      new ExnExpression(returnValue),
                                      Typ.FromTypeReference(retType),
                                      location);
+
+
             retNode.Instructions.Add(retInstr);
-            retNode.Successors = new List<CfgNode> { state.ProcDesc.ExitNode };
+            if (state.CurrentInstruction.OpCode.Code == Code.Throw)
+            {
+                var builtinFunctionExpression = new ConstExpression(ProcedureName.BuiltIn__throw);
+                var throwCall = new Call(state.GetIdentifier(Identifier.IdentKind.Normal),
+                                         new Tvoid(),
+                                         builtinFunctionExpression,
+                                         new List<Call.CallArg>(),
+                                         new Call.CallFlags(),
+                                         state.CurrentLocation);
+                retNode.Instructions.Add(throwCall);
+            }
             return retNode;
         }
 
-        protected static void CreateExceptionalEdges(ProgramState state, CfgNode entryNode)
+        /// <summary>
+        /// Gets the entry node for the exception handler, creating the entry block if needed.
+        /// flow.
+        /// </summary>
+        /// <param name="state">The state.</param>
+        /// <param name="handler">The catch handler for which we are potentially creating
+        /// exceptional control flow.</param>
+        protected static CfgNode GetHandlerExceptionalEntryBlock(ProgramState state, 
+                                                                 ExceptionHandler handler)
         {
-            foreach (var node in state.NodesToLinkWithExceptionBlock)
+            if (!(handler.HandlerType == ExceptionHandlerType.Catch || 
+                  handler.HandlerType == ExceptionHandlerType.Finally))
             {
-                node.ExceptionNodes.Add(entryNode);
+                return null;
             }
-            state.NodesToLinkWithExceptionBlock = new List<CfgNode>();
+            CfgNode entryNode;
+            Identifier exceptionIdentifier;
+            var exnInfo = state.MethodExceptionHandlers;
+            var instruction = handler.TryEnd.Previous;
+
+            if (!state.ExceptionHandlerSetToEntryNode.ContainsKey(handler))
+            {
+                if (handler.HandlerType == ExceptionHandlerType.Catch)
+                {
+                    (entryNode, exceptionIdentifier) = GetHandlerEntryNode(
+                        state, exnInfo.TryOffsetToCatchHandlers[instruction.Offset]
```

Comment:
Do we need an exception handling here in case `exnInfo.TryOffsetToCatchHandlers[instruction.Offset]` does not exist? Or why we can assume `instruction.Offset` exists in `exnInfo.TryOffsetToCatchHandlers`?

Reply:
it's guaranteed because of the                 if (handler.HandlerType == ExceptionHandlerType.Catch)